src/plonk_util.rs

```diff
@@ -77,6 +74,44 @@ pub(crate) fn halo_n<C: HaloCurve>(s_bits: &[bool]) -> C::ScalarField {
     a * C::ZETA_SCALAR + b
 }
 
+/// Compute `n(x)` for a given `x`, where `n` is the injective function related to the Halo
+/// endomorphism.
+pub(crate) fn halo_n_mul<C: HaloCurve>(s_bits: &[bool], p: AffinePoint<C>) -> AffinePoint<C> {
+    // This is based on Algorithm 1 of the Halo paper, except that we start with Acc = O.
+
+    debug_assert_eq!(s_bits.len() % 2, 0, "Number of scalar bits must be even");
+
+    let p_p = p.to_projective();
+    let p_n = p_p.neg();
+    let endo_p_p = C::endomorphism(p).to_projective();
+    let endo_p_n = endo_p_p.neg();
+
+    let mut acc = ProjectivePoint::<C>::ZERO;
+
+    for s_bits_chunk in s_bits.chunks(2) {
+        let bit_lo = s_bits_chunk[0];
+        let bit_hi = s_bits_chunk[1];
+
+        let s;
+        if bit_hi {
```

Comment:
I think this is switched from the Halo paper, where the first branch is a hi bit of 0. Doesn't really matter but I think I used that convention in the recursive version.

Reply:
Yes you're right. It was switched in the `halo_n` function, so I've changed both `halo_n` and `halo_n_mul` and it should be coherent with the paper and `CurveEndoGate` now.